runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```
@@ -129,81 +124,136 @@ public long getTail() {
     }
 
+    /**
+     * Get the first address in this bitmap.
+     * @return first address in the bitmap if its not empty, otherwise returns the trim mark
+     */
+    private long getFirst() {
+        if (bitmap.isEmpty()) {
+            return trimMark;
+        }
+
+        return bitmap.getLongIterator().next();
+    }
+
     /**
      * Add an address to this address space.
      *
      * @param address address to add.
      */
     public void addAddress(long address) {
-        //TODO(Maithem): prevent adding an address smaller than the trim mark
+        if (address <= this.trimMark || address < 0) {
+            throw new IllegalArgumentException("Address=" + address + " TrimMark=" + this.trimMark);
+        }
         bitmap.addLong(address);
     }
 
-    /**
-     * Remove addresses from the stream's address map
-     * and set the new trim mark (to the greatest of all addresses to remove).
-     */
-    private void removeAddresses(List<Long> addresses) {
-        addresses.stream().forEach(bitmap::removeLong);
-
-        // Recover allocated but unused memory
-        bitmap.trim();
-        // TODO(Maithem) if addresses is empty, the trim mark will not be updated
-        trimMark = Collections.max(addresses);
-
-        log.trace("removeAddresses: new trim mark set to {}", trimMark);
-    }
-
     /**
      * Trim all addresses lower or equal to trimMark and set new trim mark.
      *
-     * @param trimMark upper limit of addresses to trim
+     * @param newTrimMark upper limit of addresses to trim
      */
-    public void trim(long trimMark) {
-        if (!Address.isAddress(trimMark)) {
+    public void trim(long newTrimMark) {
+        if (!Address.isAddress(newTrimMark)) {
             // If not valid address return and do not attempt to trim.
-            log.warn("trim: attempting to trim non-valid address {}", trimMark);
+            log.warn("trim: attempting to trim non-valid address {}", newTrimMark);
+            return;
```

Should this throw exception instead of return?